File: 299.bulls-and-cows.py
#
# @lc app=leetcode id=299 lang=python3
#
# [299] Bulls and Cows
#

# Time:  O(n)
# Space: O(1)

from collections import defaultdict


# @lc code=start
class Solution:
    def getHint(self, secret: str, guess: str) -> str:
        lookup, a, b = defaultdict(int), 0, 0
        for s, g in zip(secret, guess):
            if s == g:
                a += 1
            else:
                b += int(lookup[s] < 0) + int(lookup[g] > 0)
                lookup[s] += 1
                lookup[g] -= 1
        return f"{a}A{b}B"


# @lc code=end

File: 299.bulls-and-cows.js
/*
 * @lc app=leetcode id=299 lang=javascript
 *
 * [299] Bulls and Cows
 */

// @lc code=start
/**
 * @param {string} secret
 * @param {string} guess
 * @return {string}
 */
var getHint = function (secret, guess) {
  const lookup = Array(10).fill(0);
  let a = 0;
  let b = 0;
  for (let i = 0; i < secret.length; ++i) {
    if (secret[i] === guess[i]) {
      ++a;
    } else {
      if (lookup[secret[i].charCodeAt() - 48]++ < 0) {
        ++b;
      }
      if (lookup[guess[i].charCodeAt() - 48]-- > 0) {
        ++b;
      }
    }
  }
  return `${a}A${b}B`;
};
// @lc code=end
